### ISSUE TITLE
performance of laplacian_matrix

### ISSUE THREAD

**Issue by dahtah** [JuliaGraphs/LightGraphs.jl#1168]
Comparing LightGraphs to igraph (a C library for graph processing), I'd noticed that computing the graph laplacian seemed quite a bit slower. The following experiment seems to bear that out. Here's my own naive attempt at forming the graph Laplacian:
```{julia }
using SparseArrays

function lap(G :: SimpleGraph)
    i = collect(1:nv(G))
    j = collect(1:nv(G))
    x = degree(G)
    for u in vertices(G)
        for v in neighbors(G,u)
            push!(i,u)
            push!(j,v)
            push!(x,-1)
        end
    end
    sparse(i,j,x)
end
```

Note that I'm not attempting to directly form a CSC representation, just passing indices to "sparse". On the following graph:
```{julia }
G = Grid([150, 150])
```
lap(G) takes 0.01 sec while laplacian_matrix(G) takes 1.3 sec.
On: 
```{julia }
G = Grid([15, 15, 15, 15])
```
I'm getting .04 sec vs 6.5 sec.
On other graphs the difference is less dramatic, for instance on: 
```{julia }
G = barabasi_albert(Int(1e4),400)
```
there's a small advantage to laplacian_matrix (for the instances I've tried), but note that the resulting matrix isn't very sparse.
Am I missing some obvious cases in which laplacian_matrix is much faster?

**Comment by simonschoelly** [comment]
From the code:
```
function laplacian_matrix(g::AbstractGraph{U}, T::DataType=Int; dir::Symbol=:unspec) where U
    if dir == :unspec
        dir = is_directed(g) ? :both : :out
    end
    A = adjacency_matrix(g, T; dir=dir)
    D = convert(SparseMatrixCSC{T,U}, Diagonal(sparse(sum(A, dims=2)[:])))
    return D - A
end
```
There are a lot of unnecessary allocations happening here  so there is definitely room for improvement if you want to work on that.

**Comment by dahtah** [comment]
All right, I'll take a look

**Comment by sbromberger** [comment]
Closed via #1191.
